File: c_and_c++_programs/cppdsaLOVEBABBARandSTRIVER/recursion/dice_combinations.cpp
#include <iostream>
#include <vector>

using namespace std;


void combinations(string processed, int target)
{
    if(target == 0)
    {
        cout<<processed<<" ";
        return ;
    }

    for(int i = 1; i<= target && i<= 6; i++)
    {
        combinations(processed+""+i, target-i);
    }
}

int main()
{
    // cout<<"absc"+2<<endl;

    combinations("", 4);


    return 0;
}

File: b01_dsa/recursion/DIceCombinations.java
import java.util.*;

public class DiceCombinations{
    static int total_combinations = 0;

    public static void dice_com(int target, int curr_sum)
    {
        // base condition
        if(target == curr_sum){
            total_combinations++;
            return;
        }
        if(curr_sum > target){
            return ;
        }

        // dice_com(target, (curr_sum + 1));
        // dice_com(target, (curr_sum + 2));
        // dice_com(target, (curr_sum + 3));
        // dice_com(target, (curr_sum + 4));
        // dice_com(target, (curr_sum + 5));
        // dice_com(target, (curr_sum + 6));

        for(int possibilities = 1; possibilities <= 6; possibilities++)
        {
            dice_com(target, (curr_sum + possibilities));
        }
    }

    public static void main(String[] args)
    {
        int target = 6;
        int curr_sum = 0;
        dice_com(target, curr_sum);

        System.out.println("total combinations = " + total_combinations);
    }

}